src/trace_comparison.rs

```diff
@@ -130,10 +121,114 @@ pub fn generate_block_comparison(
+            base_idx += 1;
+        } else {
+            // There are both base and native transactions left
+            let base_trace = &base_traces[base_idx];
+            let native_trace = &native_traces[native_idx];
+
+            let (base_tx_hash, native_tx_hash) = match (
+                base_trace["transaction_hash"].as_str(),
+                native_trace["transaction_hash"].as_str(),
+            ) {
+                (Some(base), Some(native)) => (base, native),
+                _ => panic!("Base trace and native trace are expected to be strings"),
+            };
+
+            if base_tx_hash == native_tx_hash {
+                let (base_trace, native_trace) =
+                    match (base_trace.as_object(), native_trace.as_object()) {
+                        (Some(base), Some(native)) => (base, native),
+                        _ => panic!("Base trace and native trace are expected to be objects"),
+                    };
+                compare_traces_result.push(generate_transaction_comparisons(
+                    base_trace.clone(),
+                    native_trace.clone(),
+                    base_tx_hash,
+                ));
+                base_idx += 1;
+                native_idx += 1;
+            } else if let Some(&native_match_idx) = native_hash_map.get(base_tx_hash) {
+                // Found a match for base transaction in native
+                for i in native_idx..native_match_idx {
+                    compare_traces_result.push(
+                        ComparisonResult::new_different_native_only(native_traces[i].clone())
+                            .into(),
+                    );
+                }
+                native_idx = native_match_idx;
+            } else if let Some(&base_match_idx) = base_hash_map.get(native_tx_hash) {
+                // Found a match for native transaction in base
+                for i in base_idx..base_match_idx {
+                    compare_traces_result.push(
+                        ComparisonResult::new_different_base_only(base_traces[i].clone()).into(),
+                    );
+                }
+                base_idx = base_match_idx;
+            } else {
+                // No match found, we keep adding base transactions first
+                compare_traces_result
+                    .push(ComparisonResult::new_different_base_only(base_trace.clone()).into());
+                base_idx += 1;
+            }
+        }
+    }
+
+    Value::Array(compare_traces_result)
+}
+
 /// Compares the trace root of two transactions
 /// If the trace roots are different, it compares the contract dependencies and storage dependencies.
 /// Otherwise, it skips comparison for contract dependencies and storage dependencies.
-fn generate_transaction_comparions(base_tx_trace: Value, native_tx_trace: Value) -> Value {
+fn generate_transaction_comparisons(
+    mut base_trace: Map<String, Value>,
+    mut native_trace: Map<String, Value>,
+    tx_hash: &str,
```

Comment:
I felt that it made more sense of transaction hash comparison to be done outside of this function, since it is unrelated to the main logic of comparing `"trace_root"`